ext/bson/bson_native.c

```diff
@@ -319,6 +687,71 @@ VALUE rb_bson_byte_buffer_get_string(VALUE self)
   return string;
 }
 
+
+VALUE rb_bson_byte_buffer_get_hash(VALUE self){
+  VALUE doc = Qnil;
+  byte_buffer_t *b;
+  char type;
+  VALUE cDocument = rb_const_get(rb_const_get(rb_cObject, rb_intern("BSON")), rb_intern("Document"));
+  TypedData_Get_Struct(self, byte_buffer_t, &rb_byte_buffer_data_type, b);
+
+  /* skip length */
+  ENSURE_BSON_READ(b, 4);
+  b->read_position += 4;
+  doc = rb_funcall(cDocument, rb_intern("allocate"),0);
+
+  ENSURE_BSON_READ(b, 1);
+  while((type = (uint8_t)*READ_PTR(b)) != 0){
+    VALUE field;
+    b->read_position += 1;
+    field = rb_bson_byte_buffer_get_cstring(self);
+    rb_hash_aset(doc, field, pvt_read_field(b, self, type));
+
+    ENSURE_BSON_READ(b, 1);
+  }
+  return doc;
+}
+
+VALUE rb_bson_byte_buffer_get_array(VALUE self){
+  byte_buffer_t *b;
+  VALUE array = Qnil;
+  char type;
+  TypedData_Get_Struct(self, byte_buffer_t, &rb_byte_buffer_data_type, b);
+
+  ENSURE_BSON_READ(b, 4);
+  array = rb_ary_new2(pvt_get_int32(b));
```

Comment:
The "length" of a BSON array is a document length, just like the document/hash case.  It's not the number of elements.